File: apps/api/app/schemas/admin_history.py
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict

from app.schemas.asset import SortOrder


class AdminHistorySortBy(str, Enum):
    TIMESTAMP = "timestamp"
    USER_ID = "user_id"
    METHOD = "method"
    PATH = "path"
    STATUS_CODE = "status_code"
    DURATION_MS = "duration_ms"


class AdminHistoryItemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    timestamp: datetime
    user_id: int | None
    role: str | None
    actor_email: str | None = None
    method: str
    path: str
    query: str | None
    status_code: int
    duration_ms: int
    client_ip: str | None
    user_agent: str | None
    action_name: str | None
    resource_type: str | None
    resource_id: int | None
    result: str | None
    request_body_masked: str | None
    response_body_masked: str | None


class AdminHistoryPageOut(BaseModel):
    items: list[AdminHistoryItemOut]
    total: int
    page: int
    page_size: int
    sort_by: AdminHistorySortBy
    sort_order: SortOrder

File: apps/web-asset/src/api/adminHistory.ts
import { http } from "./http";

export type AdminHistoryMethod = "GET" | "POST" | "PUT" | "PATCH" | "DELETE";

export type AdminHistoryItemOut = {
  id: number;
  timestamp: string;
  user_id: number | null;
  role: string | null;
  actor_email: string | null;
  method: string;
  path: string;
  query: string | null;
  status_code: number;
  duration_ms: number;
  client_ip: string | null;
  user_agent: string | null;
  action_name: string | null;
  resource_type: string | null;
  resource_id: number | null;
  result: string | null;
  request_body_masked: string | null;
  response_body_masked: string | null;
};

export type AdminHistoryPageOut = {
  items: AdminHistoryItemOut[];
  total: number;
  page: number;
  page_size: number;
};

export type AdminHistoryQuery = {
  from?: string;
  to?: string;
  user_id?: number;
  method?: AdminHistoryMethod;
  path_contains?: string;
  status_code?: number;
  page?: number;
  page_size?: number;
};

export async function getAdminHistory(params: AdminHistoryQuery): Promise<AdminHistoryPageOut> {
  const { data } = await http.get<AdminHistoryPageOut>("/admin/history", { params });
  return data;
}
